helix-core/src/match_brackets.rs

```diff
@@ -42,35 +46,80 @@ pub fn find_matching_bracket(syntax: &Syntax, doc: &Rope, pos: usize) -> Option<
 //
 // If no surrounding scope is found, the function returns `None`.
 #[must_use]
-pub fn find_matching_bracket_fuzzy(syntax: &Syntax, doc: &Rope, pos: usize) -> Option<usize> {
+pub fn find_matching_bracket_fuzzy(syntax: &Syntax, doc: RopeSlice, pos: usize) -> Option<usize> {
     find_pair(syntax, doc, pos, true)
 }
 
-fn find_pair(syntax: &Syntax, doc: &Rope, pos: usize, traverse_parents: bool) -> Option<usize> {
+fn find_pair(
+    syntax: &Syntax,
+    doc: RopeSlice,
+    pos_: usize,
+    traverse_parents: bool,
+) -> Option<usize> {
     let tree = syntax.tree();
-    let pos = doc.char_to_byte(pos);
+    let pos = doc.char_to_byte(pos_);
 
-    let mut node = tree.root_node().named_descendant_for_byte_range(pos, pos)?;
+    let mut node = tree.root_node().descendant_for_byte_range(pos, pos)?;
 
     loop {
-        let (start_byte, end_byte) = surrounding_bytes(doc, &node)?;
-        let (start_char, end_char) = (doc.byte_to_char(start_byte), doc.byte_to_char(end_byte));
+        if node.is_named() {
+            let (start_byte, end_byte) = surrounding_bytes(doc, &node)?;
+            let (start_char, end_char) = (doc.byte_to_char(start_byte), doc.byte_to_char(end_byte));
+
+            if is_valid_pair(doc, start_char, end_char) {
+                if end_byte == pos {
+                    return Some(start_char);
+                }
+
+                // We return the end char if the cursor is either on the start char
+                // or at some arbitrary position between start and end char.
+                if traverse_parents || start_byte == pos {
```

Comment:
I might be missing something, how is checking `traverse_parents` relevant here? We are inside a named node, we found a valid pair. Unless I'm wrong, we don't care about the value of `traverse_parents` here.

Reply:
In fact maybe we should simply always return `Some(end_char)` regardless of anything, like:
```
if end_byte == pos { return Some(start_char) } else { Some(end_char) }
```

Reply:
The check was introduced in 59e7dce33caaa57d1d16da0242accbabec7a3fef

to address https://github.com/helix-editor/helix/issues/3357#issuecomment-1411230516. When transverse_parents is false we only want to consider matches exactly on the cursor (which this check enforces)